### PR TITLE
More firebase-frameworks work

### PR DESCRIPTION
**Pull request by jamesdaniels**
* Allow source option with `emulators:start`, not just `firebase serve`
* Move the context / config manipulation into firebase-tools
* Bump the dep

### CODE REVIEW

File: src/hosting/normalizedHostingConfigs.ts [hunk 1]
```diff
@@ -90,6 +95,9 @@ export function normalizedHostingConfigs(
   cmdOptions: any, // eslint-disable-line @typescript-eslint/no-explicit-any
   options: { resolveTargets?: boolean } = {}
 ): HostingConfig[] {
+  // First see if there's a momoized copy on the options, from frameworks
```
Comment (TheIronDev) [inline src/hosting/normalizedHostingConfigs.ts:98]: nit: memoized

File: src/frameworks/index.ts [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+import { listSites, Site } from "../hosting/api";
+import { getAppConfig, AppPlatform } from "../management/apps";
+import { promises as fsPromises } from "fs";
+import { promptOnce } from "../prompt";
+
+const { writeFile } = fsPromises;
+
+export const shortSiteName = (site?: Site) => site?.name && site.name.split("/").pop();
+
+export const prepareFrameworks = async (targetNames: string[], context: any, options: any) => {
```
Comment (TheIronDev) [inline src/frameworks/index.ts:15]: Please add docs on what this function's purpose is

File: src/frameworks/index.ts [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+        if (targetNames.includes("functions")) {
+          const combinedFunctionsConfig = [functionConfig].concat(
+            options.config.get("functions") || []
+          );
+          options.config.set("functions", combinedFunctionsConfig);
+        } else {
+          targetNames.unshift("functions");
+          options.config.set("functions", functionConfig);
+        }
+      }
```
Comment (TheIronDev) [inline src/frameworks/index.ts:77]: Nit: add spaces and comments for this block of code. (L16-L77)

Reasoning - 
A lot is going on here, its hard to understand the what/why of the code when its densely packed.

File: src/frameworks/index.ts [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+      config.rewrites = [
+        ...(config.rewrites || []),
+        ...rewrites,
+        {
+          source: "**",
+          function: functionName,
+        },
+      ];
+
+      let firebaseProjectConfig = null;
```
Comment (TheIronDev) [inline src/frameworks/index.ts:88]: nit: `let firebaseProjectConfig;`

(unless null string is intentional)
(`JSON.stringify(null)` => 'null')

**Review by bkendall — APPROVED** [review]

**Pull request merged**